pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1664,58 +1646,47 @@ func (r *ReconcileClusterDeployment) mergePullSecrets(cd *hivev1.ClusterDeployme
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			cdLog.Info("Existing pull secret object not found")
-			pullSecretObjExists = false
-		} else {
-			return false, errors.Wrap(err, "Error getting pull secret from cluster deployment")
-		}
+	pullSecretSecret := &corev1.Secret{}
+	switch created, err := controllerutils.GetOrCreate(
+		context.Background(),
+		r.Client,
+		client.ObjectKey{Namespace: cd.Namespace, Name: constants.GetMergedPullSecretName(cd)},
+		pullSecretSecret,
+		func() error {
+			pullSecretSecret.Type = corev1.SecretTypeDockerConfigJson
+			pullSecretSecret.StringData = map[string]string{corev1.DockerConfigJsonKey: pullSecret}
+			return errors.Wrap(
+				controllerutil.SetControllerReference(cd, pullSecretSecret, r.scheme),
+				"error setting controller reference on new merged pull secret",
+			)
+		},
+	); {
+	case apierrors.IsConflict(err):
+		return false, err
+	case err != nil:
+		return false, errors.Wrap(err, "error getting or creating merged pull secret secret")
+	case created:
+		cdLog.WithField("secretName", pullSecretSecret.Name).Info("Created the merged pull secret object successfully")
+		return true, nil
 	}
 
-	if pullSecretObjExists {
-		existingPullSecret, ok := existingPullSecretObj.Data[corev1.DockerConfigJsonKey]
-		if !ok {
-			return false, fmt.Errorf("Pull secret %s did not contain key %s", mergedSecretName, corev1.DockerConfigJsonKey)
-		}
-		if controllerutils.GetHashOfPullSecret(string(existingPullSecret)) == controllerutils.GetHashOfPullSecret(pullSecret) {
-			cdLog.Debug("Existing and the new merged pull secret are same")
-			return false, nil
-		}
-		cdLog.Info("Existing merged pull secret hash did not match with latest merged pull secret")
-		existingPullSecretObj.Data[corev1.DockerConfigJsonKey] = []byte(pullSecret)
-		err = r.Update(context.TODO(), existingPullSecretObj)
-		if err != nil {
-			return false, errors.Wrap(err, "error updating merged pull secret object")
-		}
-		cdLog.WithField("secretName", mergedSecretName).Info("Updated the merged pull secret object successfully")
-	} else {
-
-		// create a new pull secret object
-		newPullSecretObj := generatePullSecretObj(
-			pullSecret,
-			mergedSecretName,
-			cd,
-		)
-		err = controllerutil.SetControllerReference(cd, newPullSecretObj, r.scheme)
-		if err != nil {
-			cdLog.Errorf("error setting controller reference on new merged pull secret: %v", err)
-			return false, err
-		}
-		err = r.Create(context.TODO(), newPullSecretObj)
-		if err != nil {
-			return false, errors.Wrap(err, "error creating new pull secret object")
-		}
-		cdLog.WithField("secretName", mergedSecretName).Info("Created the merged pull secret object successfully")
+	existingPullSecret, ok := pullSecretSecret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		return false, fmt.Errorf("Pull secret %s did not contain key %s", pullSecretSecret.Name, corev1.DockerConfigJsonKey)
+	}
+	if string(existingPullSecret) == pullSecret {
```

Comment:
This was changed to compare the full pull secrets instead of hashes of the pull secrets. Comparing the hashes is only helpful if we already have at least one of the hashes. In this case, we only have the full strings of both and are creating both hashes. It is more efficient, then, to simply compare the full strings.